javalin/src/main/java/io/javalin/json/JsonMapper.kt

```diff
@@ -28,6 +30,27 @@ interface JsonMapper {
      */
     fun toJsonStream(obj: Any, type: Type): InputStream = throw NotImplementedError("JsonMapper#toJsonStream not implemented")
 
+    /**
+     * Javalin uses this method for [io.javalin.http.Context.writeJsonStream]. When implementing this method
+     * you are expected to consume `stream` one element at a time, convert that element to JSON, and write that
+     * element's corresponding JSON to the `outputStream`. The intent is to reduce the amount of memory used
+     * during transformation of a data stream to JSON.
+     *
+     * It is implied that the `stream` of objects will be transformed to an Array type in JSON. In other words,
+     * it is your responsibility to surround the output with `[` and `]`.
+     *
+     * The `outputStream` will be a compressed output stream.
+     *
+     * The content type header will already be set to "application/json".
+     *
+     * Once Javalin calls this function, it will not participate any further in the response. It becomes
+     * your responsibility to deliver the remainder of the response to the `outputStream`.
+     *
+     * When your implementation returns from the `writeStream` function, assume that the underlying resources
+     * for the `stream` will be released.
+     */
+    fun writeStream(outputStream: OutputStream, stream: Stream<*>): Unit = throw NotImplementedError("JsonMapper#writeStream not implemented")
```

Comment:
Is it too silly to call this `writeStreamToStream` ?

Reply:
Asking the question made me realize that I think the arguments should be switched.

Reply:
> Is it too silly to call this writeStreamToStream ?

Yeah, that's kinda awkward to mentally map "Stream" to two completely different meanings of the word.

An unambiguous function name would be `writeStreamToOutputStream`. ??

I suppose the function name `writeStream` was just trying to say "write to an outputStream", but without saying *what* is being written. If there were ever another function to write something else to an output stream, it could keep the function name `writeStream()` but with another function signature for that type, e.g. `writeStream(outputStream: OutputStream, foo: Foo)`. But that might be overthinking for a future that cannot be foretold.

For that scenario, the function name `writeToOutputStream()` would be clear on what you are writing to, and there could be many of these functions with overloaded signatures.


> Asking the question made me realize that I think the parameters should be switched.

Sure, I can swap the args to fit Javalin's style. I just did it out of habit. I'm used to passing "the thing being acted upon" as the first arg for functions that aren't returning the result of their action. It might matter more if there were a group of functions to "act on a thing", where each function took a different numbers of arguments.

Reply:
> An unambiguous function name would be writeStreamToOutputStream. ??

I don't hate it 😅 

This is one of the most difficult to name functions I've run into in a while. I was looking at the other functions in the interface: 

```
fun toJsonString(obj: Any, type: Type)
fun toJsonStream(obj: Any, type: Type)
fun <T : Any> fromJsonString(json: String, targetType: Type): T
fun <T : Any> fromJsonStream(json: InputStream, targetType: Type): T
```

```
fun writeStream() // too ambiguous 
fun writeStreamToStream() // also ambiguous (and silly)
fun writeStreamToOutputStream() // i don't hate this
fun writeToOutputStream() // i don't hate this either
```

None of these mention "json" though, which maybe the original function didn't have to do (but they do). I don't see any way to "fix" this. 

When all is said and done, I think `writeToOutputStream` is maybe the best name.

Reply:
> None of these mention "json" though

Well, these functions *do* live inside an interface named `JsonMapper`, so a bit of "JSON" is implied. Maybe it's not so terrible.

> When all is said and done, I think writeToOutputStream is maybe the best name.

Okay, I'll update the PR. I think I'm done with writing unit tests, unless you spot something else we should test.

I also thought about looking at `JavalinGson` to see what that implementation would look like, but I could do that in another PR.

Reply:
> Well, these functions do live inside an interface named JsonMapper, so a bit of "JSON" is implied. Maybe it's not so terrible.

No, just a bit unfortunate that the other four methods included that. Not an issue.

> Okay, I'll update the PR. I think I'm done with writing unit tests, unless you spot something else we should test.

The tests look great!